cmd/commands/upgrade.go

```diff
@@ -0,0 +1,204 @@
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/urfave/cli/v2"
+	"github.com/version-fox/vfox/internal/util"
+)
+
+const SelfUpgradeName = "upgrade"
+
+var Upgrade = &cli.Command{
+	Name:   SelfUpgradeName,
+	Usage:  "upgrade vfox to the latest version",
+	Action: upgradeCmd,
+}
+
+func fetchLatestVersion() (string, error) {
+	response, err := http.Get("https://github.com/version-fox/vfox/tags")
+	if err != nil {
+		return "", err
+	}
+	defer response.Body.Close()
+
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		return "", err
+	}
+	re, err := regexp.Compile(`href="/version-fox/vfox/releases/tag/(v[0-9.]+)"`)
+	if err != nil {
+		return "", err
+	}
+	matches := re.FindAllStringSubmatch(string(body), -1)
+
+	if len(matches) == 0 {
+		return "", fmt.Errorf("Failed to fetch the version.")
+	}
+
+	latestVersion := matches[0][1]
+	return latestVersion, nil
+}
+
+func constructBinaryName(tagName string) string {
+	osType := strings.ToLower(runtime.GOOS)
+	if osType == "darwin" {
+		osType = "macos"
+	}
+
+	archType := runtime.GOARCH
+	if archType == "arm64" {
+		archType = "aarch64"
+	}
+	if archType == "amd64" {
+		archType = "x86_64"
+	}
+
+	extName := "tar.gz"
+	if osType == "windows" {
+		extName = "zip"
+	}
+
+	fileName := fmt.Sprintf("vfox_%s_%s_%s.%s", tagName[1:], osType, archType, extName)
+	return fileName
+}
+
+func generateUrls(currVersion string, tagName string) (string, string) {
+	fileName := constructBinaryName(tagName)
+	binURL := fmt.Sprintf("https://github.com/version-fox/vfox/releases/download/%s/%s", tagName, fileName)
+	diffURL := fmt.Sprintf("https://github.com/version-fox/vfox/compare/%s...%s", currVersion, tagName)
+	return binURL, diffURL
+}
+
+func downloadFile(filepath string, url string) error {
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	_, err = io.Copy(out, resp.Body)
+	return err
+}
+
+func upgradeCmd(ctx *cli.Context) error {
+	currVersion := fmt.Sprintf("v%s", ctx.App.Version)
+	latestVersion, err := fetchLatestVersion()
+	if err != nil {
+		return cli.Exit("Failed to fetch the latest version: "+err.Error(), 1)
+	}
+	fmt.Println("Current version: ", currVersion)
+	fmt.Println("Latest available:", latestVersion)
+
+	if currVersion == latestVersion {
+		return cli.Exit("vfox is already up to date.", 0)
+	}
+	if err = RequestPermission(); err != nil {
+		return err
+	}
+	exePath, err := os.Executable()
+	if err != nil {
+		return cli.Exit("Failed to get executable path: "+err.Error(), 1)
+	}
+	exeDir, exeName := filepath.Split(exePath)
+
+	binURL, diffURL := generateUrls(currVersion, latestVersion)
+	tempFile := "vfox_latest.tar.gz"
+	if runtime.GOOS == "windows" {
+		tempFile = "vfox_latest.zip"
+	}
+	tempFile = filepath.Join(os.TempDir(), tempFile)
+	tempDir := filepath.Join(os.TempDir(), "vfox_upgrade")
```

Comment:
Use `VFOX_HOME/tmp`
See #156

Reply:
`VFOX_HOME` may also be on a different file system.

Reply:
All right.